### Import libraries

In [1]:
# Import libraries
import pandas as pd
import numpy as np
import os

### Import Data

In [2]:
# Import data
path = r"C:\Users\Juliana\Documents\DATA ANALYTICS COURSE\IMMERSION\4. Python\Instacart Basket Analysis"
df_prods = pd.read_csv(os.path.join(path,'02 Data','Original Data','products.csv'), index_col = False)
df_ords = pd.read_csv(os.path.join(path,'02 Data','Prepared Data','orders_wrangled.csv'), index_col = False)

### 02. Descriptive Statistics

### a. Descriptive statistics for df_prods

In [3]:
df_prods.describe()

product_id      aisle_id  department_id        prices
count  49693.000000  49693.000000   49693.000000  49693.000000
mean   24844.345139     67.770249      11.728433      9.994136
std    14343.717401     38.316774       5.850282    453.519686
min        1.000000      1.000000       1.000000      1.000000
25%    12423.000000     35.000000       7.000000      4.100000
50%    24845.000000     69.000000      13.000000      7.100000
75%    37265.000000    100.000000      17.000000     11.200000
max    49688.000000    134.000000      21.000000  99999.000000

##### The maximum value for the "Price" column doesn't appear to be correct. Therefore, further investigation is required.

In [6]:
# Find products with high unusual prices
df_prods.loc[df_prods['prices']> 30]

product_id                      product_name  aisle_id  department_id  \
21554       21553  Lowfat 2% Milkfat Cottage Cheese       108             16   
33666       33664             2 % Reduced Fat  Milk        84             16   

        prices  
21554  14900.0  
33666  99999.0

###### Both of these prices appear to have been entered incorrectly, and need to be corrected by getting the accurate prices from Instacart. However, for the purposes of this exercise I will change them to 1.49 and 9.99 respectively.

In [5]:
df_prods=df_prods.replace({'prices':{14900.0:1.49,99999.0:9.99}})

In [6]:
# Check the new descriptive statistics
df_prods.describe()

product_id      aisle_id  department_id        prices
count  49693.000000  49693.000000   49693.000000  49693.000000
mean   24844.345139     67.770249      11.728433      7.682190
std    14343.717401     38.316774       5.850282      4.200262
min        1.000000      1.000000       1.000000      1.000000
25%    12423.000000     35.000000       7.000000      4.100000
50%    24845.000000     69.000000      13.000000      7.100000
75%    37265.000000    100.000000      17.000000     11.200000
max    49688.000000    134.000000      21.000000     25.000000

### b. Descriptive statistics for df_ords

In [7]:
df_ords.describe()

Unnamed: 0      order_id       user_id  order_number  \
count  3.421083e+06  3.421083e+06  3.421083e+06  3.421083e+06   
mean   1.710541e+06  1.710542e+06  1.029782e+05  1.715486e+01   
std    9.875817e+05  9.875817e+05  5.953372e+04  1.773316e+01   
min    0.000000e+00  1.000000e+00  1.000000e+00  1.000000e+00   
25%    8.552705e+05  8.552715e+05  5.139400e+04  5.000000e+00   
50%    1.710541e+06  1.710542e+06  1.026890e+05  1.100000e+01   
75%    2.565812e+06  2.565812e+06  1.543850e+05  2.300000e+01   
max    3.421082e+06  3.421083e+06  2.062090e+05  1.000000e+02   

       orders_day_of_week  order_hour_of_day  days_since_last_order  
count        3.421083e+06       3.421083e+06           3.214874e+06  
mean         2.776219e+00       1.345202e+01           1.111484e+01  
std          2.046829e+00       4.226088e+00           9.206737e+00  
min          0.000000e+00       0.000000e+00           0.000000e+00  
25%          1.000000e+00       1.000000e+01           4.000000e+00  
50%          3.000000e+00       1.300000e+01           7.000000e+00  
75%          5.000000e+00       1.600000e+01           1.500000e+01  
max          6.000000e+00       2.300000e+01           3.000000e+01

#### Obervations: There is nothing in the data that appears to be anomalous or requires further investigation. The minimum and maximum values in each column align with the expected range of values for each variable.

### 03. Check for mixed-data types

#### a. Check for mixed types for df_prods

In [8]:
# Check for mixed types for df_prods
for col in df_prods.columns.tolist():
  weird = (df_prods[[col]].applymap(type) != df_prods[[col]].iloc[0].apply(type)).any(axis = 1)
  if len (df_prods[weird]) > 0:
    print (col)

product_name


#### Observations: The column product_name from pd_prods has mixed-type data

#### b. Check for mixed types for df_ords

In [9]:
# Check for mixed types for df_ords
for col in df_ords.columns.tolist():
  weird = (df_ords[[col]].applymap(type) != df_ords[[col]].iloc[0].apply(type)).any(axis = 1)
  if len (df_ords[weird]) > 0:
    print (col)

#### Observations: There are not columns with mixed-type data in df_ords

### 04. Fix the mixed-type data

In [10]:
# Fix the mixed-type columns in df_prods
df_prods['product_name'] = df_prods['product_name'].astype('str')

### 05 06. Missing values

#### a. Finding missing values in pf_prods

In [11]:
# Finding missing values in pf_prods
df_prods.isnull().sum()

product_id       0
product_name     0
aisle_id         0
department_id    0
prices           0
dtype: int64

In [12]:
# Create a new dataframe
df_nan=df_prods[df_prods['product_name'].isnull()== True]

In [13]:
df_nan

Empty DataFrame
Columns: [product_id, product_name, aisle_id, department_id, prices]
Index: []

In [14]:
# compare the number of rows in your current dataframe with the number in your subset
df_prods.shape

(49693, 5)

In [15]:
# Create new dataframe without the missing values
df_prods_clean=df_prods[df_prods['product_name'].isnull()==False]

In [16]:
df_prods_clean.shape

(49693, 5)

##### Observations:In the 'df_prods' dataset, 16 missing values were found in the 'product_name' column. Since this variable does not contain numerical values, it is not possible to assign arbitrary values to fill the missing entries. Therefore, to maintain data integrity, the 16 rows corresponding to the missing values have been removed from the dataset.

#### b. Finding missing values in pf_ords

In [17]:
# Finding missing values in pf_ords
df_ords.isnull().sum()

Unnamed: 0                    0
order_id                      0
user_id                       0
eval_set                      0
order_number                  0
orders_day_of_week            0
order_hour_of_day             0
days_since_last_order    206209
dtype: int64

In [18]:
# Create a new dataframe of missing values
df_ords_nan=df_ords[df_ords['days_since_last_order'].isnull()== True]

In [19]:
df_ords_nan

Unnamed: 0  order_id  user_id eval_set  order_number  \
0                 0   2539329        1    prior             1   
11               11   2168274        2    prior             1   
26               26   1374495        3    prior             1   
39               39   3343014        4    prior             1   
45               45   2717275        5    prior             1   
...             ...       ...      ...      ...           ...   
3420930     3420930    969311   206205    prior             1   
3420934     3420934   3189322   206206    prior             1   
3421002     3421002   2166133   206207    prior             1   
3421019     3421019   2227043   206208    prior             1   
3421069     3421069   3154581   206209    prior             1   

         orders_day_of_week  order_hour_of_day  days_since_last_order  
0                         2                  8                    NaN  
11                        2                 11                    NaN  
26                        1                 14                    NaN  
39                        6                 11                    NaN  
45                        3                 12                    NaN  
...                     ...                ...                    ...  
3420930                   4                 12                    NaN  
3420934                   3                 18                    NaN  
3421002                   6                 19                    NaN  
3421019                   1                 15                    NaN  
3421069                   3                 11                    NaN  

[206209 rows x 8 columns]

##### Observation: The number of missing values: 206,209, in the 'days_since_last_order' column coincides with the number of unique customers (user_id). Additionally, all of these missing values have an 'order_number' of 1, which indicates that they correspond to the first order from each customer. Since it wouldn't make sense to impute any other value in this scenario, I will leave the missing values as they are. Analysts may utilize the order number (1) as a flag to identify  that it is the first transaction of each customer.

### 07 08. Duplicates

#### a. Find duplicates for df_prods_clean

In [20]:
# Find duplicates
df_dups= df_prods_clean[df_prods_clean.duplicated()]

In [21]:
df_dups

product_id                                       product_name  \
462           462                  Fiber 4g Gummy Dietary Supplement   
18459       18458                                         Ranger IPA   
26810       26808               Black House Coffee Roasty Stout Beer   
35309       35306  Gluten Free Organic Peanut Butter & Chocolate ...   
35495       35491                            Adore Forever Body Wash   

       aisle_id  department_id  prices  
462          70             11     4.8  
18459        27              5     9.2  
26810        27              5    13.4  
35309       121             14     6.8  
35495       127             11     9.9

In [22]:
# Addressing duplicates
df_prods_clean.shape

(49693, 5)

In [23]:
df_prods_clean_no_dups=df_prods_clean.drop_duplicates()

In [24]:
df_prods_clean_no_dups.shape

(49688, 5)

##### Observations:  The duplicated found have been deleted

#### b. Find duplicates for df_ords

In [25]:
# Find duplicates
df_ords_dups= df_ords[df_ords.duplicated()]

In [26]:
df_ords_dups

Empty DataFrame
Columns: [Unnamed: 0, order_id, user_id, eval_set, order_number, orders_day_of_week, order_hour_of_day, days_since_last_order]
Index: []

##### Observations: There are no duplicates

### 09. Export final, cleaned data

In [27]:
df_prods_clean_no_dups.to_csv(os.path.join(path, '02 Data','Prepared Data', 'products_checked.csv'))

In [28]:
df_ords.to_csv(os.path.join(path, '02 Data','Prepared Data','orders_checked.csv'))